### ISSUE TITLE
Ability to always rerun a repo rule

### ISSUE THREAD

**Issue by kchodorow** [bazelbuild/bazel#3041]
I could have sworn that the `stamp` attribute combined with `--stamp` and `--workspace_status_command` did this, but the rule isn't being run even though `bazel-out/volatile-status.txt` changes. See https://groups.google.com/d/msgid/bazel-discuss/c1a4f036-19e6-45e1-b1eb-e6c280f3c6a5%40googlegroups.com.

**Comment by lberki** [comment]
I don't think we should implement this. Bazel builds are supposed to be reproducible and it's very easy to introduce an "always run" genrule somewhere deep in your dependency graph, thus making everything that depends on it build slowly.

We also don't want to make Bazel a general purpose workflow system, so I'd much rather keep the invariant that inputs are files, full stop.

**Comment by lberki** [comment]
@kchodorow : we actually special- case `volatile-status.txt` so that if that's the only input of an action that's changed, said action is not re-run.

**Comment by philwo** [comment]
> We also don't want to make Bazel a general purpose workflow system

Who decides that? I personally would like to be able to use Bazel as a general purpose workflow system, especially when this is the only thing in the way of having that use case work.

I don't think it's kind to close this without further discussion.

**Comment by c-parsons** [comment]
Drive-by comment: Apple systems are somewhat notorious for hiding system information outside of the normal filesystem. It would be nice if our current local cc-osx and xcode repository rules could be invalidated based on the output of a script. Right now we recommend a `bazel clean` which is a very poor alternative.

This is at least something we should consider.

**Comment by lberki** [comment]
I thought that has been decided since a good while ago; that's why we don't have the corresponding feature internally. I obviously can't link to that discussion here (I also tried to look for it but couldn't seem to find it), but that's why we don't have unconditionally evaluated genrules there, either.

We also had a thread about it on bazel-discuss (at that time, I was in the opposite corner): https://groups.google.com/forum/#!msg/bazel-discuss/L40qovu-d1s/P08-N2xZDgAJ

I haven't given a lot of deep thought to the use case of @c-parsons but it seems like it'd better be encapsulated as a repository rule. After all, that's where we capture non-hermeticity.

Routing this to core folk for another look.

**Comment by kchodorow** [comment]
A repository rule doesn't actually help @c-parsons's case, as it still will be cached build-to-build (there's no file that changes we can depend on).

**Comment by c-parsons** [comment]
I admit that allowing us to have an always-run repository rule would be fine for my use-case (allowing always-run non-repository-rule functionality would be unnecessary for me)

**Comment by ericfelly** [comment]
Nathan can you take a look at this? Thanks.

**Comment by excavador** [comment]
I definitely need always-run rule

**Comment by haxorz** [comment]
[i hadn't set up email filters for bazel github emails until last night, so i'm only seeing this github issue now]

@lberki - what's being asked of me and/or the core team here? are you looking for another opinion on this FR? or are you looking for a statement on the viability of implementing this FR?

i think this FR is fine, as long as the documentation is very clear that this feature should be very rarely used and will be hostile to performance. i think this sort of FR is in the same category as bazel's current support for arbitrary external symlinks, which is to say "it's correct, but we don't care about optimizing this use case because we don't encourage this use case".

as for the implementation, the existing "ErrorTransienceValue" in skyframe could be generalized to the aptly named "TransienceValue", and make the ActionExecutionValue for an action from rule have a direct dep on the TransienceValue (and also have the action have a unique & random action cache key) and then everything works for free. but here's where the performance "concern" above comes in here: on incremental builds, we will unconditionally invalidate the UTC of the TransienceValue. down the line, we'll rerun the action. skyframe change pruning will still come into play for the output and UTC of this output, but at non-zero cost.

**Comment by excavador** [comment]
@haxorz sounds great.

I need something like for that for following:
- inspect actual docker images
- inspect actual docker container
- inspect actual process list.

I need all these things:
- for writing build rules - for instance, I have rule which transform bunch of sql files to YAML-file. It implemented using docker (for postgresql) and pyrseas (third-party tool which inspect postgresql information schema and generate yaml file).
- for writing tests - for I should setup bunch of docker containers and processes for testing. I, obviously CAN implement it as bazel said, but it works weird - about 30-40 seconds just for build postgresql pg_data with database schema BEFORE I can do anything.
- for writing run targets - for wake-up the local backend stand, which used by frontend team for integration.

Yes, I understand all penalties, why it is wrong, etcetera. But it is the last part to keep bazel-only for all tasks related to build & test.

**Comment by haxorz** [comment]
assigning back to @lberki for prioritization and re-assignment

**Comment by lberki** [comment]
@haxorz: I wanted to get an opinion from you folks about exposing the "always execute this action" bit to Skylark. I think it's inevitable, but it's a powerful tool and I'm not sure the use cases we know of warrant the inclusion of such a footgun in Bazel.

In particular, I am worried about two things:

* That people will use the "always run" bit as a way to get away with neglecting declaring dependencies correctly
* That such actions will creep into every build thus slowing them down

@dslomov , what do you think? IMHO we should do it, but if we do do it, we'll eventually need diagnostics to figure out which actions during a particular build were run because they were tagged as "always run".

**Comment by haxorz** [comment]
if we add this feature, i could imagine banning it internally in google (or only adding it to bazel in the first place). i am not aware of any compelling use cases internally (all of the use cases in this github issue and in the linked bazel-discuss thread involve doing non-hermetic things) and the risks (poisoning builds with slowness) are large, especially in a mono-repo where the incentives for caring about build performance aren't aligned correctly.

i agree about the diagnostics part. internally, the distributor INFO log that is currently written is very useful in general and would also be useful for this. externally, the best we have right now (that i know of) is the bazel INFO log.

**Comment by excavador** [comment]
```
In particular, I am worried about two things:

That people will use the "always run" bit as a way to get away with neglecting declaring dependencies correctly
That such actions will creep into every build thus slowing them down
```

Why do not disable it by default, and enable only through bazelrc?
It would be perfectly save for beginner users, and power users which definitely need it always can enable it.

**Comment by excavador** [comment]
Right now I forced to write wrapper-script "tools/bzl" like following
```
#!/bin/bash -eu
ROOT="$(dirname $(dirname $(realpath ${BASH_SOURCE[0]})))"
MARKER="${ROOT}/bzl.marker"
date +%s > "${MARKER}"
bazel $@
```
and always run `tools/bzl`  instead of bazel.

In my build files/skylark code I add dependency from these files where I need it.
It is not very usefull

**Comment by excavador** [comment]
Moreover, full script is following
```
#!/bin/bash -eu

ROOT="$(dirname $(realpath ${BASH_SOURCE[0]}))"
MARKER="${ROOT}/bzl.marker"
VERSION_REQUIRED="0.7.0"

function update_marker() {
    local
    date +%s > "${MARKER}"
}

function check_bazel() {
    local INSTALLED="$(which bazel | wc -l | awk '{print $1}')"
    if [[ "${INSTALLED}" == "0" ]]; then
        echo "you need to install bazel (brew install bazel)"
        exit 1
    fi
    local VERSION_ACTUAL="$(bazel version | grep label | awk '{print $3}' | sed -e 's:-homebrew$::g')"
    if [[ "${VERSION_REQUIRED}" != "${VERSION_ACTUAL}" ]]; then
        echo "wrong version bazel (actual ${VERSION_ACTUAL} required ${VERSION_REQUIRED})"
        exit 1
    fi
}

update_marker
check_bazel
bazel $@
```

**Comment by aiuto** [comment]
I was almost going to close this out because I am not convinced it is needed, but the discussion is a good one for posterity.
Moving to the product and starlark teams to think about the merits of this.

My alternate strawman design:
This is deliberately hard for an arbitrary rule to use it. That prevents random foot-gun.
- bazel can read a list of paths from <WORKSPACE_TOP>/.bazel_volatile_paths
- the file system interface treats those files as always changing. The time stamp and content for each is different on each call.
- we can use visibility on those files to lock down their use.
- A CI system could swap the volatile paths file before the build to get some level of determinism.
- No starlark semantic changes.

**Comment by golvok** [comment]
_Just to add another possible use-case, though I kinda agree that any breaks from the ideal of ideal of repoducibility should possibly be avoided._

While transitioning to bazel from Make, it is sometimes easier to have `local = 1` genrules with side-efffects that copy files into the source tree where the old Makefiles expect them. However, since these are side-effects (ie. not an output of the rule), bazel will not re-run the rule if the in-tree file gets deleted or something. For us, it would be sufficient to be able to do something like `bazel run-it-anyway //path/to:genrule` to fixup the tree.

This can actually be achieved by using another genrule to produce an executable script, which is then run by the original genrule. (`bazel run` will work or that script). However, you lose access to some features (eg. 'make' variable substitutions) and may have to double-escape quotes, etc.

**Comment by brandjon** [comment]
Maybe I misunderstand this thread, but thought we already had a concept of repo rules that always run.

Is this FR only related to repository rules, and not the rest of the build (where we really should be hermetic)? If so we should deprioritize as repo rules will be deprecated by bzlmod. (Triaging under that basis.)

**Comment by keithl-stripe** [comment]
Isn't [the `no-cache` tag](https://docs.bazel.build/versions/master/be/common-definitions.html#common.tags) designed to do just this? Does that tag not always work?

**Comment by burdiyan** [comment]
@keithl-stripe I think it doesn't help here. The `no-cache` tag disables disk-based or remote caching, but still will only re-run the rule if inputs have changed.

**Comment by uri-canva** [comment]
I agree that "always rerun repository rule" doesn't sound like a well defined feature, and the example use cases in this thread are very different, and some even seem to relate to rules, not repository rules. I though I'd summarise some of the existing bazel features that can help in some situations where one might want this feature:

1. `--workspace_status_command` lets you pass any input you'd like into a rule, assuming you can serialise it. You can even generate files then pass the sha and path to the files you generated. Stable keys are considered inputs that affect the cache so you can use it to declare a lot of inputs that aren't handled by the usual bazel input mechanisms. One possible improvement here is for rules to be able to declare which values of the status output they depend on, right now changing any of the stable values will cause all rules that use any of the values to be rebuilt.
2. `environ` in `repository_rule` lets you declare which environment variables the repository rule depends on, and will re-evaluate it if it changes. You can use this to pass inputs to your repository rules, or use it to pass some sha / timestamp the rule can use to track whether some input it uses has changed or not. You can generate this environment variable value in a `tools/bazel` wrapper.
3. `managed_directories` allows you to declare directories in the source tree that can be managed by external processes (there's not much documentation for it, but see the original proposal https://docs.google.com/document/d/1uaeQpPrSH4q46zGuXLgtzMkLVXk7Er9I-P_-ScwwGLk/edit#heading=h.xgjl2srtytjt).

**Comment by blorente** [comment]
At the risk of reviving a dead thread, I _think_ we have a use case for "run on every bazel invocation" repository rules, as detailed in #17640:

We'd like the ability to re-run a repository rule every time your Xcode path changes (every time the result of `xcode-select -p` changes), so that we can configure their build accordingly.

These changes are so frequent (in some cases) and the cost of checking so low that we would be willing to pay the cost of always re-running them at every bazel invocation.

We work with a variety of teams, not all of whom can provide access to their code. We have several limitations to implement @uri-canva 's suggestions:
- We have limited control over their bazel installation, which limits the usefulness of the `tools/bazel` wrapper approach to pass the value of `xcode-select -p` as an input through `environ`. We would like to avoid mandating a bazel wrapper.
- In macOS, The actual value of `xcode-select -p` is stored in a part of the machine that would require disabling system integrity protection to read. We would like to avoid imposing that on all users, ruling out the `managed_directories` approach. Furthermore, as far as I know, `managed_directories` is going away anyway.

Currently, the best we can offer is a troubleshooting tool that will check that their Xcode installation is the same as the one bazel thinks it is, but it would be very nice to have things just work.

**Comment by hanikesn** [comment]
To add another use-case:
I just migrated our container image builds from `rules_docker` to `rules_oci` using their [stamp_tags](https://github.com/bazel-contrib/rules_oci/blob/main/examples/push/stamp_tags.bzl) example to stamp our push rules with a unique tag based on the hash of the current state of the git repo to make sure we always have a unique name, when triggering a (test-) deploy locally. But caching is now too good that the tag isn't repushed all the time.

**Comment by bcsgh** [comment]
Yet another use case: I'm generating a chain of x509 test certs that expire and need to be regenerated every day or so, even if the code uses to generate them hasn't changed. I could use `bazel clean` but I'd rather something that's a bit more surgical and can be used more much more often than actually needed and without causing *everything* to rebuild.

What I'm currently playing with is:

```
_BUILD = "exports_files({files})"

def _status_repository_impl(ctx):
    VERSIONS = {
      "year":   "+%Y",
      "month":  "+%Y-%m",
      "week":   "+%Gw%V",
      "day":    "+%Y-%m-%d",
      "hour":   "+%Y-%m-%d %H",
      "minute": "+%Y-%m-%d %H:%M",
      "second": "+%Y-%m-%d %H:%M:%S",
    }

    date = ctx.which("date")
    DATES = dict([
      (i, ctx.execute([date, "--universal", f]))
      for i,f in VERSIONS.items()
    ])

    err = dict([
      (i, f.stderr.strip())
      for i,f in DATES.items()
      if f.return_code
    ])
    if err: fail(err)

    for f, d in DATES.items(): ctx.file(f, d.stdout.strip())
    ctx.file("BUILD", _BUILD.format(files=str(sorted(DATES.keys()))))
    return

volatile_repository = repository_rule(
    doc = """Create a repository with files that change at predictable time.

        NOTE: Bazel makes it hard to get this to re-run every time
        because that could very easily become a foot-gun. See:
        https://github.com/bazelbuild/bazel/issues/3041

        To force the regeneration, run:
        $ bazel sync --configure
    """,
    implementation = _status_repository_impl,
    local = True,
    configure = True,
)
```

**Comment by burdiyan** [comment]
@bcsgh If you can afford to use a wrapper script for Bazel (which you could place in `tools/bazel` in your repo, to make Bazel automatically use it), the easiest way I've found to achieve something like this is to set an environment variable inside the wrapper script, with a random value on every invocation, and make the repository rule depend on that environment variable.

Be careful though if any of your build actions use default shell env, i.e. make sure you use `--incompatible_strict_action_env` flag, otherwise you'll invalidate lots of build actions unnecessarily.

**Comment by quic-torgil** [comment]
I'm also hit by this issue, trying to forward some runtime information to a `bazel run` invocation. Since my project have downstream users I'm avoiding all workarounds involving wrapping bazel.

@uri-canva
> --workspace_status_command lets you pass any input you'd like into a rule

Yes. And with the existence of a key in stable-status.txt I can (at build/run time) verify that a script generating that
key is run and error out for features needing this. Downsides is that it requires that a script is run under 
`--workspace_status_command` and that you don't have the information at analyze time.

I managed to rerun a repo rule consistently by letting it depend on `@//:bazel-out/stable-status.txt` and update
that with volatile information but since that file is generated after the repo rule I cannot guarantee that i've been run until build time where I can verify validity through a client PID or similar.

@bcsgh
>  volatile_repository = repository_rule

Any success with this?  I needed a `bazel shutdown` in order to rerun such repository rule.
